src/cargo/ops/registry.rs

```diff
@@ -90,7 +91,22 @@ pub struct PublishOpts<'cfg> {
 
 pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
     let specs = opts.to_publish.to_package_id_specs(ws)?;
+
+    if let Packages(_) = &opts.to_publish {
+        if specs.len() > 1 {
+            bail!("the `-p` argument must be specified to select a single package to publish");
+        }
```

Comment:
Is there a particular reason to have this check within the `if let`?  Could it perhaps be done just after the `to_package_id_specs`?  And I don't think the `is_virtual` needs to check the `to_publish` variant.

Reply:
Because i find 
```
pub enum Packages {
    Default,
    All,
    OptOut(Vec<String>),
    Packages(Vec<String>),
}
```
I want to express if when we meet `-p`  we should make these check. And in other enum we should make other check. Maybe it make code verbose. I know now we only have `-p` and `Default` argument. But if we add the other two argument in the future one day, we can add more check in different case clearly.

Reply:
To the second, if user `cargo -p SPEC` in a virtual workspace. We should let it run. So i think we need check `is_virtual` in `to_publish` variant. Exactly your first suggestion:

> In the root of a virtual workspace, it should be an error to run without -p.

Reply:
I think the ultimate goal is to only allow one package to be selected, irrespective of which `Packages` variant is used. There are a few issues with the code as-is:

This code will panic in the following situations:

```toml
# Root Cargo.toml
[workspace]
members = ["bar"]
default-members = ["bar"]
[package]
name = "foo"
version = "0.1.0"
```

Here, `to_publish` is `Packages::Default`, but there are multiple packages returned from `members_with_features` (it is a peculiarity of how the old member resolution works). I think because of this, the assert should be just a normal error (`if pkgs.len() > 1 {bail!(…)}`).

----

Similarly, but for different reasons:

```toml
# Root Cargo.toml
[workspace]
members = ["bar"]
default-members = ["bar", "bar2"]
resolver = "2"
[package]
name = "foo"
version = "0.1.0"
```

Here the length of `specs` is 2, but the check is being circumvented.  I think the specs length check should not care what variant of `Packages` is used.

Reply:
Ok. I updated. To the second, i find the reason of it is below, it will add the "current" package when we use `old member resolution`. So i think we can filter `pkgs` to get right `pkg`. And i cover more cases which you mentioned.
https://github.com/rust-lang/cargo/blob/39ad1039d9e3e1746177bf5d134af4c164f95528/src/cargo/core/workspace.rs#L1450-L1458